In [1]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib as plt
import plotly.express as px

In [3]:
SHG = pd.read_csv('SHG_Booking.csv')

SHG.head(20)

Booking ID   Hotel    Booking Date   Arrival Date  Lead Time  Nights  \
0            1  Resort  7/24/2014 0:00  7/1/2015 0:00        342       0   
1            2  Resort  6/24/2013 0:00  7/1/2015 0:00        737       0   
2            3  Resort  6/24/2015 0:00  7/1/2015 0:00          7       1   
3            4  Resort  6/18/2015 0:00  7/1/2015 0:00         13       1   
4            5  Resort  6/17/2015 0:00  7/1/2015 0:00         14       2   
5            6  Resort  6/17/2015 0:00  7/1/2015 0:00         14       2   
6            7  Resort   7/1/2015 0:00  7/1/2015 0:00          0       2   
7            8  Resort  6/22/2015 0:00  7/1/2015 0:00          9       2   
8            9  Resort   4/7/2015 0:00  7/1/2015 0:00         85       3   
9           10  Resort  4/17/2015 0:00  7/1/2015 0:00         75       3   
10          11  Resort   6/8/2015 0:00  7/1/2015 0:00         23       4   
11          12  Resort  5/27/2015 0:00  7/1/2015 0:00         35       4   
12          13  Resort  4/24/2015 0:00  7/1/2015 0:00         68       4   
13          14  Resort  6/13/2015 0:00  7/1/2015 0:00         18       4   
14          15  Resort  5/25/2015 0:00  7/1/2015 0:00         37       4   
15          16  Resort  4/24/2015 0:00  7/1/2015 0:00         68       4   
16          17  Resort  5/25/2015 0:00  7/1/2015 0:00         37       4   
17          18  Resort  6/19/2015 0:00  7/1/2015 0:00         12       1   
18          19  Resort   7/1/2015 0:00  7/1/2015 0:00          0       1   
19          20  Resort  6/24/2015 0:00  7/1/2015 0:00          7       4   

    Guests  Distribution Channel Customer Type         Country Deposit Type  \
0        2                Direct     Transient        Portugal   No Deposit   
1        2                Direct     Transient        Portugal   No Deposit   
2        1                Direct     Transient  United Kingdom   No Deposit   
3        1             Corporate     Transient  United Kingdom   No Deposit   
4        2   Online Travel Agent     Transient  United Kingdom   No Deposit   
5        2   Online Travel Agent     Transient  United Kingdom   No Deposit   
6        2                Direct     Transient        Portugal   No Deposit   
7        2                Direct     Transient        Portugal   No Deposit   
8        2   Online Travel Agent     Transient        Portugal   No Deposit   
9        2  Offline Travel Agent     Transient        Portugal   No Deposit   
10       2   Online Travel Agent     Transient        Portugal   No Deposit   
11       2   Online Travel Agent     Transient        Portugal   No Deposit   
12       2   Online Travel Agent     Transient   United States   No Deposit   
13       3   Online Travel Agent     Transient           Spain   No Deposit   
14       2   Online Travel Agent     Transient        Portugal   No Deposit   
15       2   Online Travel Agent     Transient         Ireland   No Deposit   
16       2  Offline Travel Agent      Contract        Portugal   No Deposit   
17       2   Online Travel Agent     Transient         Ireland   No Deposit   
18       2             Corporate     Transient          France   No Deposit   
19       2                Direct     Transient  United Kingdom   No Deposit   

    Avg Daily Rate     Status   Status Update  Cancelled (0/1)  Revenue  \
0                0  Check-Out   7/1/2015 0:00                0        0   
1                0  Check-Out   7/1/2015 0:00                0        0   
2               75  Check-Out   7/2/2015 0:00                0       75   
3               75  Check-Out   7/2/2015 0:00                0       75   
4               98  Check-Out   7/3/2015 0:00                0      196   
5               98  Check-Out   7/3/2015 0:00                0      196   
6              107  Check-Out   7/3/2015 0:00                0      214   
7              103  Check-Out   7/3/2015 0:00                0      206   
8               82   Canceled   5/6/2015 0:00                1        0 

In [4]:
f' There are {SHG.shape[0]} rows and {SHG.shape[1]} columns'

' There are 119390 rows and 17 columns'

In [5]:
# .info to return the data type, null values and other important description

SHG.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   Booking ID            119390 non-null  int64 
 1   Hotel                 119390 non-null  object
 2   Booking Date          119390 non-null  object
 3   Arrival Date          119390 non-null  object
 4   Lead Time             119390 non-null  int64 
 5   Nights                119390 non-null  int64 
 6   Guests                119390 non-null  int64 
 7   Distribution Channel  119390 non-null  object
 8   Customer Type         119390 non-null  object
 9   Country               118902 non-null  object
 10  Deposit Type          119390 non-null  object
 11  Avg Daily Rate        119390 non-null  int64 
 12  Status                119390 non-null  object
 13  Status Update         119390 non-null  object
 14  Cancelled (0/1)       119390 non-null  int64 
 15  Revenue          

In [6]:
# .isnull for inspecting missing values

SHG.isna().sum()

Booking ID                0
Hotel                     0
Booking Date              0
Arrival Date              0
Lead Time                 0
Nights                    0
Guests                    0
Distribution Channel      0
Customer Type             0
Country                 488
Deposit Type              0
Avg Daily Rate            0
Status                    0
Status Update             0
Cancelled (0/1)           0
Revenue                   0
Revenue Loss              0
dtype: int64

# 1)since my geographical analysis heavily depends on it, i filled my missing values with "Unknown"

In [9]:
SHG['Country'].fillna('Unknown', inplace=True)

# 2) Converted my Date Column to Data Time because they are currently stored as object
# (i.e Just plain text or string format) which limit what i can do with them


In [11]:
SHG['Booking Date'] = pd.to_datetime(SHG['Booking Date'])
SHG['Arrival Date'] = pd.to_datetime(SHG['Arrival Date'])
SHG['Status Update'] = pd.to_datetime(SHG['Status Update'])

In [12]:
SHG

Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
0                1  Resort   2014-07-24   2015-07-01        342       0   
1                2  Resort   2013-06-24   2015-07-01        737       0   
2                3  Resort   2015-06-24   2015-07-01          7       1   
3                4  Resort   2015-06-18   2015-07-01         13       1   
4                5  Resort   2015-06-17   2015-07-01         14       2   
...            ...     ...          ...          ...        ...     ...   
119385      119386    City   2017-08-07   2017-08-30         23       7   
119386      119387    City   2017-05-21   2017-08-31        102       7   
119387      119388    City   2017-07-28   2017-08-31         34       7   
119388      119389    City   2017-05-14   2017-08-31        109       7   
119389      119390    City   2017-02-05   2017-08-29        205       9   

        Guests  Distribution Channel Customer Type         Country  \
0            2                Direct     Transient        Portugal   
1            2                Direct     Transient        Portugal   
2            1                Direct     Transient  United Kingdom   
3            1             Corporate     Transient  United Kingdom   
4            2   Online Travel Agent     Transient  United Kingdom   
...        ...                   ...           ...             ...   
119385       2  Offline Travel Agent     Transient         Belgium   
119386       3   Online Travel Agent     Transient          France   
119387       2   Online Travel Agent     Transient         Germany   
119388       2   Online Travel Agent     Transient  United Kingdom   
119389       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status Status Update  Cancelled (0/1)  \
0        No Deposit               0  Check-Out    2015-07-01                0   
1        No Deposit               0  Check-Out    2015-07-01                0   
2        No Deposit              75  Check-Out    2015-07-02                0   
3        No Deposit              75  Check-Out    2015-07-02                0   
4        No Deposit              98  Check-Out    2015-07-03                0   
...             ...             ...        ...           ...              ...   
119385   No Deposit              96  Check-Out    2017-09-06                0   
119386   No Deposit             225  Check-Out    2017-09-07                0   
119387   No Deposit             158  Check-Out    2017-09-07                0   
119388   No Deposit             104  Check-Out    2017-09-07                0   
119389   No Deposit             151  Check-Out    2017-09-07                0   

        Revenue  Revenue Loss  
0             0             0  
1             0             0  
2            75             0  
3            75             0  
4           196             0  
...         ...           ...  
119385      673             0  
119386     1578             0  
119387     1104             0  
119388      731             0  
119389     1361             0  

[119390 rows x 17 columns]

# 3) I Checked for Duplicates
# To Ensure there are no repeated records.

In [16]:
SHG.duplicated().sum()
SHG.drop_duplicates(inplace=True)

# 4) Validating and Fixing my Data Types

## a) converted my Booking ID → string
### Why: Because i'm not performing math on it, Treating it as a string prevents accidental numerical operations.

In [19]:
SHG['Booking ID'] = SHG['Booking ID'].astype(str)

## b) Converted Avg Daily Rate, Revenue, Revenue Loss → float
### Why: These are monetary values. Keeping them as int64 might hide decimals For price-related analysis, float ensures precision

In [21]:
SHG['Avg Daily Rate'] = SHG['Avg Daily Rate'].astype(float)
SHG['Revenue'] = SHG['Revenue'].astype(float)
SHG['Revenue Loss'] = SHG['Revenue Loss'].astype(float)

## c) Lead Time, Nights, Guests → int is OK
### Why: These are count-based metrics. Integer is appropriate.

#### I Checked: Ensure no negative values:

In [22]:
SHG[['Lead Time', 'Nights', 'Guests']].describe()

Lead Time         Nights         Guests
count  119390.000000  119390.000000  119390.000000
mean      104.011416       3.427900       1.968239
std       106.863097       2.557439       0.722394
min         0.000000       0.000000       0.000000
25%        18.000000       2.000000       2.000000
50%        69.000000       3.000000       2.000000
75%       160.000000       4.000000       2.000000
max       737.000000      69.000000      55.000000

## d) Converted Hotel, Distribution Channel, Customer Type, Deposit Type, Country, Status → category

### Why: These are qualitative variables with repeated values. Using category:

#### By converting these columns to category, i'm explicitly telling pandas (and other people):

#### "These columns represent categories — not free-form text or numerical data."
##### This helps:

##### Prevent errors in analysis.

In [23]:
categorical_cols = ['Hotel', 'Distribution Channel', 'Customer Type', 'Deposit Type', 'Country', 'Status']
SHG[categorical_cols] = SHG[categorical_cols].astype('category')


### 5) "Validating Business Logic"
#### checking if the data makes sense based on real-world rules or expectations.

#### a) If a booking is labeled "Canceled", it should have the binary cancellation flag set to 1.

###### Checked for mismatch

In [25]:
SHG[SHG['Status'] == 'Canceled'][['Status', 'Cancelled (0/1)']].value_counts()

Status    Cancelled (0/1)
Canceled  1                  43017
Name: count, dtype: int64

#### b) If Revenue = 0, is Revenue Loss > 0?
##### If the hotel made no revenue, it could mean:
###### The guest canceled last minute → so you lost money (Revenue Loss > 0).
###### Or the booking was valid but the room was free (Revenue Loss = 0).

In [26]:
SHG[SHG['Revenue'] == 0][['Revenue Loss', 'Status']].value_counts()

Revenue Loss  Status   
 0.0          Check-Out    1746
-248.0        Canceled      429
-124.0        Canceled      286
-120.0        Canceled      274
-126.0        Canceled      246
                           ... 
-406.0        No-Show         1
-1486.0       Canceled        1
-1487.0       Canceled        1
-402.0        No-Show         1
-6300.0       Canceled        1
Name: count, Length: 2368, dtype: int64

#### c) Are there rows where Guests == 0?
##### You can't book a room with zero guests (usually).

In [27]:
SHG[SHG['Guests'] == 0]

Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
2224         2225  Resort   2015-10-05   2015-10-06          1       3   
2409         2410  Resort   2015-10-12   2015-10-12          0       0   
3181         3182  Resort   2015-10-15   2015-11-20         36       3   
3684         3685  Resort   2015-07-18   2015-12-30        165       5   
3708         3709  Resort   2015-07-18   2015-12-30        165       6   
...           ...     ...          ...          ...        ...     ...   
115029     115030    City   2017-03-12   2017-06-27        107       3   
115091     115092    City   2017-06-29   2017-06-30          1       1   
116251     116252    City   2017-06-01   2017-07-15         44       2   
116534     116535    City   2017-07-13   2017-07-15          2       7   
117087     117088    City   2017-02-07   2017-07-27        170       2   

        Guests  Distribution Channel    Customer Type             Country  \
2224         0             Corporate  Transient-Party            Portugal   
2409         0             Corporate        Transient            Portugal   
3181         0   Online Travel Agent  Transient-Party               Spain   
3684         0   Online Travel Agent  Transient-Party            Portugal   
3708         0   Online Travel Agent  Transient-Party            Portugal   
...        ...                   ...              ...                 ...   
115029       0   Online Travel Agent        Transient         Switzerland   
115091       0                Direct        Transient            Portugal   
116251       0   Online Travel Agent        Transient              Sweden   
116534       0   Online Travel Agent  Transient-Party  Russian Federation   
117087       0  Offline Travel Agent        Transient              Brazil   

       Deposit Type  Avg Daily Rate     Status Status Update  Cancelled (0/1)  \
2224     No Deposit             0.0  Check-Out    2015-10-06                0   
2409     No Deposit             0.0  Check-Out    2015-10-12                0   
3181     No Deposit             0.0  Check-Out    2015-11-23                0   
3684     No Deposit             0.0  Check-Out    2016-01-04                0   
3708     No Deposit             0.0  Check-Out    2016-01-05                0   
...             ...             ...        ...           ...              ...   
115029   No Deposit           101.0  Check-Out    2017-06-30                0   
115091   No Deposit             0.0  Check-Out    2017-07-01                0   
116251   No Deposit            74.0  Check-Out    2017-07-17                0   
116534   No Deposit            23.0  Check-Out    2017-07-22                0   
117087   No Deposit             0.0  Check-Out    2017-07-29                0   

        Revenue  Revenue Loss  
2224        0.0           0.0  
2409        0.0           0.0  
3181        0.0           0.0  
3684        0.0           0.0  
3708        0.0           0.0  
...         ...           ...  
115029    302.0           0.0  
115091      0.0           0.0  
116251    148.0           0.0  
116534    160.0           0.0  
117087      0.0           0.0  

[180 rows x 17 columns]

In [28]:
SHG[SHG['Guests'] == 0]['Status'].value_counts()


Status
Check-Out    155
Canceled      24
No-Show        1
Name: count, dtype: int64

In [29]:
SHG[SHG['Guests'] == 0]['Cancelled (0/1)'].value_counts()


Cancelled (0/1)
0    155
1     25
Name: count, dtype: int64

In [30]:
SHG[SHG['Guests'] == 0][['Revenue', 'Revenue Loss']].describe()


Revenue  Revenue Loss
count  180.000000         180.0
mean    51.933333           0.0
std    145.293023           0.0
min      0.000000           0.0
25%      0.000000           0.0
50%      0.000000           0.0
75%      0.000000           0.0
max    819.000000           0.0

# found out there are illogical reasonings 

* checked out bookings with zero guests
* some have non-zero revenue - which should not happen

## i dropped it completely because 
* they don't follow business logic
* they are few (180 out of 119,390) that's 0.15%
* they could mislead my analysis

In [31]:
SHG = SHG[SHG['Guests'] != 0]

In [32]:
SHG.to_csv('cleaned_SHG_Data.csv', index=False)